includes/amp-helper-functions.php

```diff
@@ -464,38 +464,43 @@ function amp_register_default_scripts( $wp_scripts ) {
 		]
 	);
 
-	// Get all AMP components as defined in the spec.
-	$extensions = [];
+	// Register all AMP components as defined in the spec.
 	foreach ( AMP_Allowed_Tags_Generated::get_allowed_tag( 'script' ) as $script_spec ) {
-		if ( isset( $script_spec[ AMP_Rule_Spec::TAG_SPEC ]['extension_spec']['name'], $script_spec[ AMP_Rule_Spec::TAG_SPEC ]['extension_spec']['version'] ) ) {
-			$versions = $script_spec[ AMP_Rule_Spec::TAG_SPEC ]['extension_spec']['version'];
-			array_pop( $versions );
-			$extensions[ $script_spec[ AMP_Rule_Spec::TAG_SPEC ]['extension_spec']['name'] ] = array_pop( $versions );
+		if ( ! isset( $script_spec[ AMP_Rule_Spec::TAG_SPEC ]['extension_spec'] ) ) {
+			continue;
 		}
-	}
-
-	if ( isset( $extensions['amp-carousel'] ) ) {
-		/*
-		 * The 0.2 version of amp-carousel depends on the amp-base-carousel component, but this is still experimental.
-		 * Also, the validator spec does not currently specify what base dependencies a given component has.
-		 * @todo Revisit once amp-base-carousel is no longer experimental. Add support for obtaining a list of extensions that depend on other extensions to include in the script dependencies when registering below.
-		 */
-		$extensions['amp-carousel'] = '0.1';
-	}
+		$extension_spec = $script_spec[ AMP_Rule_Spec::TAG_SPEC ]['extension_spec'];
+		$versions       = array_filter(
+			$extension_spec['version'],
+			function ( $version ) {
+				// @todo Why are we including the latest version in the generated spec in the first place? We can just include the highest number version.
+				return 'latest' !== $version;
+			}
+		);
 
-	foreach ( $extensions as $extension => $version ) {
 		$src = sprintf(
 			'https://cdn.ampproject.org/v0/%s-%s.js',
-			$extension,
-			$version
+			$extension_spec['name'],
+			array_pop( $versions )
 		);
 
 		$wp_scripts->add(
-			$extension,
+			$extension_spec['name'],
 			$src,
 			[ 'amp-runtime' ],
 			null
 		);
+
+		$wp_scripts->add_data( $extension_spec['name'], 'amp_requires_usage', ! empty( $extension_spec['requires_usage'] ) );
```

Comment:
I'm not super happy with what I did here. The post-processor shouldn't be looking at the `wp_scripts()` object to determine later if this requires usage. It would be better if the `extension_specs` were obtained again during post-processing and the `required_usage` was obtained at that point.

Reply:
How about even moving the entire registration to a point where we know what is needed and what is not? That would avoid registering scripts we don't need, and you could have all checks against the spec in one place...

Reply:
These scripts are needing to be registered in `WP_Scripts` to ensure that they will be available to enqueue for AMP components used outside of AMP pages. For example, Newspack is using `amp-carousel` in its Carousel block and the block will `wp_enqueue_script( 'amp-carousel' )` which will ensure it works in AMP pages and non-AMP pages alike. (Granted, this is not guaranteed to work prior to Bento AMP, but still, it works most of the time and is better than a blank element.)